### PR TITLE
add(position): Adds position to CtTypeReference

### PR DESCRIPTION
**Pull request by tdurieux**
TODO: 

* [x] Tests

### CODE REVIEW

File: src/main/java/spoon/support/compiler/jdt/JDTTreeBuilder.java [hunk 1]
```diff
@@ -688,6 +688,26 @@ public CtPackageReference getPackageReference(PackageBinding reference) {
 
 		final Map<TypeBinding, CtTypeReference> bindingCache = new HashMap<>();
 
+		public <T> CtTypeReference<T> getTypeReference(TypeReference[] references, TypeReference ref) {
```
Comment (GerardPaligot) [inline src/main/java/spoon/support/compiler/jdt/JDTTreeBuilder.java]: What is the array of references? Can you add Javadoc or rename the parameter?

File: src/main/java/spoon/support/compiler/jdt/JDTTreeBuilder.java [hunk 2]
```diff
@@ -688,6 +688,26 @@ public CtPackageReference getPackageReference(PackageBinding reference) {
...
+					output.add(ctRef);
+				}
+			}
+			if (output.isEmpty()) {
+				return getTypeReference(ref);
+			}
+			if (output.size() == 1) {
+				return (CtTypeReference<T>) output.get(0);
+			}
+			return factory.Type().<T>createIntersectionTypeReference(output);
```
Comment (GerardPaligot) [inline src/main/java/spoon/support/compiler/jdt/JDTTreeBuilder.java]: This shouldn't be the default return.

File: src/main/java/spoon/support/compiler/jdt/JDTTreeBuilder.java [hunk 3]
```diff
@@ -775,6 +802,9 @@ private boolean isCorrectTypeReference(TypeReference ref) {
 					? factory.Package().getOrCreate(concatSubArray(namesParameterized, index)).getReference()
 					: factory.Package().topLevel();
 			inner.setPackage(packageReference);
+
+			setPositionReference(res, ref);
```
Comment (GerardPaligot) [inline src/main/java/spoon/support/compiler/jdt/JDTTreeBuilder.java]: Why do you have this invocation here and [there](https://github.com/INRIA/spoon/pull/718/files#diff-072d4b8ceab9c4e090f675ba9342b6f9R779)?

File: src/main/java/spoon/support/compiler/jdt/JDTTreeBuilder.java [hunk 4]
```diff
@@ -1255,6 +1285,13 @@ public JDTTreeBuilder(Factory factory) {
 		}
 		CtTypeReference<?> currentReference = typeReference;
 
+		setPositionReference(typeReference, type);
+		int sourceStartDeclaration = type.sourceStart();
+		int sourceStartSource = type.sourceStart();
+		int sourceEnd = type.sourceEnd();
+		CompilationUnit cu = factory.CompilationUnit().create(new String(context.compilationunitdeclaration.getFileName()));
+		typeReference.setPosition(factory.Core().createSourcePosition(cu, sourceStartDeclaration, sourceStartSource, sourceEnd, context.compilationunitdeclaration.compilationResult.lineSeparatorPositions));
```
Comment (GerardPaligot) [inline src/main/java/spoon/support/compiler/jdt/JDTTreeBuilder.java]: You can't put the reference in the context to do that in the `JDTTreeBuilder#enter` method?

File: src/main/java/spoon/support/compiler/jdt/JDTTreeBuilder.java (hunk 2 above)
Comment (tdurieux) [inline src/main/java/spoon/support/compiler/jdt/JDTTreeBuilder.java]: It is not the default expression, it is only the last one.

File: src/main/java/spoon/support/compiler/jdt/JDTTreeBuilder.java (hunk 1 above)
Comment (tdurieux) [inline src/main/java/spoon/support/compiler/jdt/JDTTreeBuilder.java]: > Can you add JavaDoc or rename the parameter?
No really, I copy the behaviour from the orignal one, but I do not know the specification of the method. 
Please, adds the JavaDoc to orignal one and I will adapt it.

File: src/main/java/spoon/support/compiler/jdt/JDTTreeBuilder.java (hunk 4 above)
Comment (tdurieux) [inline src/main/java/spoon/support/compiler/jdt/JDTTreeBuilder.java]: No because we do not  `JDTTreeBuilder#enter` all typeReferences

**Comment by tdurieux** [comment]
@GerardPaligot  can you reproduce the failing MainTest.testTest on your hardware?
I test is passing on mine.

**Pull request closed without merging**